Web Todo(React) step3 (서버환경과 기능추가)

cocoa-salon opened the pull request:
- 할일 삭제 외에 완료 기능을 추가로 구현하였습니다.
- 삭제, 완료에 따른 업데이트는 하나의 메소드에서 처리하여 중복을 줄이고자 하였습니다.
- 특정 컴포넌트 내에서만 사용되는 스타일은 해당 컴포넌트 내에서 정의하였습니다. 

감사합니다.

Commit:
2.리팩토링

- 하위 컴포넌트 별도 파일로 분리

Commit:
3.리팩토링

- state의 todoList를 JSX 문자열로부터 분리
- JSX 문자열 생성은 render 메소드 내에 위치

Commit:
4.[express](http://4.express) 환경 구축(myapp)

- todolist url 요청 및 JSON 리턴

Commit:
5 기능추가, 리팩토링

- 컴포넌트 클래스 내 생성자 메소드 화살표 함수 변경
- 할일 삭제 외 완료 처리 기능 구현
- 할일 상태에 따른 각 할일 개수 UI 구현

crongro commented on myapp/app.js:
```diff
@@ -0,0 +1,41 @@
+const app = express();
+const indexRouter = require('./routes/index');
+
+app.all('/*', function(req, res, next) {
```
모든 url에 불필요하게 header설정을 해줄필요는 없을 거 같아요.
필요한 url요청에 대해서만 cors설정을 해주세요. 

그리고 express에서 쉽게 cors를 해주는 방법도 있긴할거에요.

crongro commented on myapp/app.js:
```diff
@@ -0,0 +1,41 @@
+
+// catch 404 and forward to error handler
+app.use(function(req, res, next) {
+  next(createError(404));
```
404 잘 되는지 한번 테스트해보세요.

cocoa-salon replied:
지정된 접두어 외에 다른 키워드로 요청시 404 에러를 발생시킵니다.

crongro commented on myapp/routes/index.js:
```diff
@@ -0,0 +1,10 @@
+const todoList = require('../public/javascripts/todoList');
+
+/* GET todoList. */
+router.get('/getTodoList', function(req, res, next) {
```
router분리해서 잘 구현했어요.

crongro commented on src/App_AddTodo_AddTodoButton.js:
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
```
파일네이밍이 좀 어색합니다.
Add가 두번이나 들어가기도 하고요.
react 다른 프로젝트 네이밍도 살펴보시고, 언더바를  많이 쓰는지도 한번 들여다보세요.

crongro commented on src/App_TodoListView_DoneTodoButton.js:
```diff
@@ -0,0 +1,18 @@
+import React from 'react';
```
```
App/
    components/
         Todolist/
             mainview.jsx
             doneTodoButton.js
```

이런체계도 좋을 거 같네요.

crongro commented on src/App_TodoScore.js:
```diff
@@ -0,0 +1,37 @@
+import styled, { css } from 'styled-components'
+import { styles } from './styles'
+
+function TodoScore(props) {
```
destructuring해서 쓰면 좀더 깔끔할 듯.

crongro reviewed (changes requested):
구두로 말씀드린 부분위주로 수정하시면 좋을 거 같아요.

Commit:
6.리팩토링

- 컴포넌트 js파일을 UI 계층에 따라 폴더로 분리
- 할일 상태에 따른 각각의 state list를 todoList 하나로 통합
- 불필요한 인스턴스 변수 삭제

Commit:
7.리팩토링

- npm CORS 설치 후 특정 요청에 한해 cors 허용하도록 설정.
- 스타일 컴포넌트 import시 디스트럭처링 활용하여 변수를 간소화.
- 모든 컴포넌트 App 디렉토리를 중심으로 재위치, 컴포넌트 이름 변경 및 간소화.

crongro commented on src/App/components/TodoMain/TodoMain.js:
```diff
@@ -0,0 +1,125 @@
+
+    render() {
+        // 실제 할일만 렌더링 
+        const todoLiElement = this.state.todoList.reduce((accumulator, todo) => {
```
잘했어요. 이부분도 할일이 많아지면 역시 함수로 분리가능하겠죠.

crongro commented on src/App/components/TodoMain/TodoList.js:
```diff
@@ -0,0 +1,22 @@
+import { DoneTodoButton } from './TodoView/DoneTodoButton'
+
+// 할일 리스트 li 컴포넌트
+function TodoList(props) {
```
function component  활용을 잘했네요.

crongro reviewed (approved):
컴포넌트 분리가 잘 됐네요!
잘했습니다.
